### PR TITLE
swarm/network: Correct neighborhood depth

### PR DESCRIPTION
**Pull request by nolash**
This PR cherry-picks and amends fix by @zelig to neighborhood depth calculation from branch `ethersphere/push-syncing-pss`.

Before this fix, the depth would always be at the first proxpeer.

This fix makes sure that if there are vacant bins between the nearest non-prox peer and the first proxpeer, the first vacant bin will be the depth. It adjusts necessary tests accordingly.

It also removes number literals in kademlia params initialization.

### CODE REVIEW

**Comment by nonsense** [comment]
I tried to run the `TestNeighborhoodDepth` on this branch and on `master`, but when I run the tests from `swarm/network/kademlia_test.go` on my machine, they panic.

**Comment by nolash** [comment]
@nonsense I can't see this panic on Travis, and I can't reproduce locally. Do you have a stack trace please?

File: swarm/network/kademlia.go [hunk 1]
```diff
@@ -89,6 +99,7 @@ type Kademlia struct {
 	nDepth     int      // stores the last neighbourhood depth
 	nDepthC    chan int // returned by DepthC function to signal neighbourhood depth change
 	addrCountC chan int // returned by AddrCountC function to signal peer count change
+	Pof        func(pot.Val, pot.Val, int) (int, bool)
```
Comment (nonsense) [inline swarm/network/kademlia.go]: Missing comment on function.

File: swarm/network/kademlia.go [hunk 2]
```diff
@@ -29,6 +29,16 @@ import (
 	"github.com/ethereum/go-ethereum/swarm/pot"
 )
 
+const (
```
Comment (janos) [inline swarm/network/kademlia.go]: Any reason for creating this constants? I see no grater code clarity with them separated from NewKadParams, just an additional indirection when observing the code, and no abstraction introduced.

File: swarm/network/kademlia.go (hunk 1 above)
Comment (janos) [inline swarm/network/kademlia.go]: What is the point of exposing the Pof field in Kademlia struct, and it is set to a global variable pof in NewKademlia function. Is Kademlia.Pof needed outside of this package? If it is needed, why exposing it as a struct field if it is always the same as package global variable? I do not see that Kademlia.Pof is used anywhere in this PR or in elsewhere in the current codebase.

Comment (nolash) [reply swarm/network/kademlia.go]: It's used for the PSS neighbourhood calculations, which is a different PR.

File: swarm/network/kademlia.go (hunk 2 above)
Comment (nolash) [reply swarm/network/kademlia.go]: I'm just allergic to literals in constructors, that's all.

**Comment by nolash** [comment]
> Is there any insight what is causing the flakiness?

My _guess_ is that the healthy-checks are engineered to work with the previous (but faulty) kademlia calculations as source of truth. Now that the kademlia calculation is _correct_ (which evaluation is the object of this PR) the tests relying on these calculations have to be re-evaluated. Before that happens, it's likely that dependent tests will flake. We can leave them, but then Travis may fail and we can't merge (or?)

File: swarm/network/kademlia_test.go [hunk 3]
```diff
@@ -73,6 +76,69 @@ func Register(k *Kademlia, regs ...string) {
...
+
+	kad.Off(fartherPeer)
+	log.Trace(kad.string())
+	time.Sleep(time.Millisecond)
+	depth = kad.NeighbourhoodDepth()
+	if depth != 0 {
+		t.Fatalf("expected depth 0, was %d", depth)
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (nolash) [reply swarm/network/kademlia_test.go]: :bowing_man:

File: swarm/network/kademlia.go (hunk 2 above)
Comment (janos) [reply swarm/network/kademlia.go]: Good to know. ;) I am not allergic to unneeded indirections, but I do not like them very much.

File: swarm/network/kademlia.go (hunk 1 above)
Comment (janos) [reply swarm/network/kademlia.go]: Shouldn't be this change only be relevant to a different PR only? I think that code hygiene is nice to have when merging to master. What if this implementation changes in a different PR?

**Comment by janos** [comment]
> > Is there any insight what is causing the flakiness?
> 
> My _guess_ is that the healthy-checks are engineered to work with the previous (but faulty) kademlia calculations as source of truth. Now that the kademlia calculation is _correct_ (which evaluation is the object of this PR) the tests relying on these calculations have to be re-evaluated. Before that happens, it's likely that dependent tests will flake. We can leave them, but then Travis may fail and we can't merge (or?)

As @holisticode said, other tests are relying on this calculations. This means that they will be flaky also. Should we disable them, too?

**Comment by nolash** [comment]
> As @holisticode said, other tests are relying on this calculations. This means that they will be flaky also. Should we disable them, too?

@janos Hm. Well. Some of them already flake, I guess, and maybe even _because_ the calculations were incorrect. I don't have a good answer. And I don't know which ones all of htem are.

**Comment by janos** [comment]
> > As @holisticode said, other tests are relying on this calculations. This means that they will be flaky also. Should we disable them, too?
> 
> @janos Hm. Well. Some of them already flake, I guess, and maybe even _because_ the calculations were incorrect. I don't have a good answer. And I don't know which ones all of htem are.

It is easy to identify the test that need to be disabled, all of them that are using the tested functionality of already disabled test: `Simulation.WaitTillHealthy` function. A simple grep will do.

If your guess is that it would be the best to disable this test and leave others, from experience while working on this PR, then that is fine for me. But please, comment why the test is disabled as it would be a great help in the future.

**Comment by nolash** [comment]
> It is easy to identify the test that need to be disabled, all of them that are using the tested functionality of already disabled test: Simulation.WaitTillHealthy function. A simple grep will do.

Yes, of course. However, I meant of _all_ tests that somehow depend on the wrong neighborhood depth calc. That said, I'll put the cart back behind the horse and see what flakes come out.

**Comment by janos** [comment]
Thanks. For skipping TestWaitTillHealthy, just a simple comment with some context would be much appreciated.

File: swarm/network/kademlia.go (hunk 2 above)
Comment (nolash) [reply swarm/network/kademlia.go]: Too bad there are no macros in go :/

File: swarm/network/simulation/kademlia_test.go [hunk 4]
```diff
@@ -28,6 +28,8 @@ import (
 )
 
 func TestWaitTillHealthy(t *testing.T) {
+
+	t.Skip("temporarily disabled as simuations.WaitTillHealthy cannot be trusted")
```
Comment (holisticode) [inline swarm/network/simulation/kademlia_test.go]: minor nitpicking: typo in text (simuations)

Comment (nolash) [reply swarm/network/simulation/kademlia_test.go]: copy/paste with typo <3

**Comment by nolash** [comment]
From gitter:

> I disagree that "temporarily disabled as simulations.WaitTillHealthy cannot be trusted" is adding any context to the reasons why the test is disabled.
> comment should help in future investigation about the root cause, and there is no context about it
> but that is not important as usually we have such comments anyway in the codebase
no need to change anything, I just want to raise the awareness

@janos Noted and I agree that it should have explicitly named the cause. Thanks.

**Comment by janos** [comment]
> @janos Noted and I agree that it should have explicitly named the cause. Thanks.

Thank you, @nolash.

**Pull request merged**